golang/testsuite/networks_impl/test_network.go

```diff
@@ -19,54 +19,53 @@ import (
 )
 
 //  A custom Network implementation is intended to make test-writing easier by wrapping low-level
 //    NetworkContext calls with custom higher-level business logic
 type TestNetwork struct {
-	networkCtx            *networks.NetworkContext
-	datastoreServiceImage string
-	apiServiceImage       string
-	datastoreService      *datastore.DatastoreService
-	personModifyingApiService *api.ApiService
-	personRetrievingApiService *api.ApiService
-	nextApiServiceId      int
+	networkCtx                *networks.NetworkContext
+	datastoreServiceImage     string
+	apiServiceImage           string
+	datastoreIPAddress        string
+	datastorePort             int
+	personModifyingApiClient  *api_service_client.APIClient
+	personRetrievingApiClient *api_service_client.APIClient
+	nextApiServiceId          int
 }
 
 func NewTestNetwork(networkCtx *networks.NetworkContext, datastoreServiceImage string, apiServiceImage string) *TestNetwork {
 	return &TestNetwork{
-		networkCtx:            networkCtx,
-		datastoreServiceImage: datastoreServiceImage,
-		apiServiceImage:       apiServiceImage,
-		datastoreService:      nil,
-		personModifyingApiService: nil,
-		personRetrievingApiService: nil,
-		nextApiServiceId:      0,
+		networkCtx:                networkCtx,
+		datastoreServiceImage:     datastoreServiceImage,
+		apiServiceImage:           apiServiceImage,
+		datastoreIPAddress:        "",
+		datastorePort:             0,
+		personModifyingApiClient:  nil,
+		personRetrievingApiClient: nil,
+		nextApiServiceId:          0,
 	}
 }
 
 //  Custom network implementations usually have a "setup" method (possibly parameterized) that is used
 //   in the Test.Setup function of each test
 func (network *TestNetwork) SetupDatastoreAndTwoApis() error {
-	if network.datastoreService != nil {
-		return stacktrace.NewError("Cannot add datastore service to network; datastore already exists!")
-	}
-	if network.personModifyingApiService != nil || network.personRetrievingApiService != nil {
+
+	if network.personModifyingApiClient != nil || network.personRetrievingApiClient != nil {
 		return stacktrace.NewError("Cannot add API services to network; one or more API services already exists")
 	}
 
 	configFactory := datastore.NewDatastoreContainerConfigFactory(network.datastoreServiceImage)
-	uncastedDatastore, hostPortBindings, _, err := network.networkCtx.AddService(datastoreServiceId, configFactory)
+	datastoreServiceContext, hostPortBindings, err := network.networkCtx.AddService(datastoreServiceId, configFactory)
 	if err != nil {
 		return stacktrace.Propagate(err, "An error occurred adding the datastore service")
 	}
 
-	castedDatastore := uncastedDatastore.(*datastore.DatastoreService)
-	datastoreClient := datastore_service_client.NewDatastoreClient(castedDatastore.GetServiceContext().GetIPAddress(), castedDatastore.GetPort())
+	datastoreClient := datastore_service_client.NewDatastoreClient(datastoreServiceContext.GetIPAddress(), datastore.Port)
```

Comment:
Oh my god this is so much better than doing the service downcasting..

Reply:
yes!! I fill the same when I was doing the changes